### PR TITLE
Unit test refactor

### PR DESCRIPTION
**Pull request by gilbertlee-amd**
Unit test refactor
- This overhauls the entire unit test system by providing a simple to use framework for current and future tests that supports both single and multi process RCCL testing
- Additionally allows for improved test filtering by targeting min/max GPU ranks to test on, datatypes, reduction ops, etc via environment variables
- Modified test cases to adjust coverage across multiple tests instead of via combinatorial explosion - speeding up testing
- Supports pre-mult scalars and customized data preparation functions.
- Single executable now supports both single-process and multi-process testing

### CODE REVIEW

File: test/common/PrepDataFuncs.cpp [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+      ERROR("Number of output elements (%lu) exceeds the number of allocated output elements (%lu)\n",
+            collArgs.numOutputElements, collArgs.numOutputElementsAllocated);
+      return TEST_FAIL;
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode DefaultPrepData_Broadcast(CollectiveArgs &collArgs)
+  {
+    if (CheckAllocation(collArgs) != TEST_SUCCESS) return TEST_FAIL;
```
Comment (akolliasAMD) [inline test/common/PrepDataFuncs.cpp:53]: Simple `CHECK_CALL(CheckAllocation(collArgs))` should work here instead right? I assume that the extra line: `ERROR("Error in call %s\n", #func);  ` in check call wont be a problem in this case

File: test/common/PrepDataFuncs.cpp [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    // Otherwise all other ranks expected output is the same as input of root
+    return collArgs.expected.FillPattern(collArgs.dataType,
+                                         collArgs.numInputElements,
+                                         collArgs.root,
+                                         false);
+  }
+
+  ErrCode DefaultPrepData_Reduce(CollectiveArgs &collArgs, bool const isAllReduce)
+  {
+    if (CheckAllocation(collArgs) != TEST_SUCCESS) return TEST_FAIL;
```
Comment (akolliasAMD) [inline test/common/PrepDataFuncs.cpp:80]: `CHECK_CALL(CheckAllocation(collArgs)) ` same as above

File: test/common/PrepDataFuncs.cpp [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    if (collArgs.redOp == ncclAvg && (isAllReduce || collArgs.root == collArgs.globalRank))
+    {
+      CHECK_CALL(result.DivideByInt(collArgs.dataType, collArgs.numInputElements, collArgs.totalRanks));
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode DefaultPrepData_Gather(CollectiveArgs &collArgs, bool const isAllGather)
+  {
+    if (CheckAllocation(collArgs) != TEST_SUCCESS) return TEST_FAIL;
```
Comment (akolliasAMD) [inline test/common/PrepDataFuncs.cpp:150]: `CHECK_CALL(CheckAllocation(collArgs)) ` same as above (next time it happens again too)

File: test/common/PtrUnion.cpp [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+    {
+      free(this->ptr);
+      this->ptr = nullptr;
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::ClearGpuMem(size_t const numBytes)
+  {
+    if (hipMemset(this->ptr, 0, numBytes) != hipSuccess)
```
Comment (akolliasAMD) [inline test/common/PtrUnion.cpp:97]: could replace if statement here with `CHECK_HIP(hipMemset(this->ptr, 0, numBytes))` for consistency through the tests

File: test/common/PtrUnion.cpp [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+      {
+        if (hipMallocManaged(&I1, numBytes) != hipSuccess)
+        {
+          ERROR("Unable to allocate managed memory of GPU memory (%lu bytes)\n", numBytes);
+          return TEST_FAIL;
+        }
+      }
+      else
+      {
+        if (hipMalloc(&I1, numBytes) != hipSuccess)
```
Comment (akolliasAMD) [inline test/common/PtrUnion.cpp:51]: ccould replace if statement here with `CHECK_HIP(hipMalloc(&I1, numBytes))` for consistency through the tests

File: test/common/PtrUnion.cpp [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    {
+      int    valueI = (globalRank + i) % 256;
+      double valueF = 1.0L/((double)valueI+1.0L);
+      temp.Set(dataType, i, valueI, valueF);
+    }
+
+    // If this is GPU memory, copy from CPU temp buffer
+    if (isGpuMem)
+    {
+      if (hipMemcpy(this->ptr, temp.ptr, numBytes, hipMemcpyHostToDevice) != hipSuccess)
```
Comment (akolliasAMD) [inline test/common/PtrUnion.cpp:135]: could replace if statement here with CHECK_HIP(hipMemset(this->ptr, 0, numBytes)) for consistency through the tests
that would also need a comment here with the error stated underneath though

File: test/common/TestBed.hpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    // Execute all collectives on all test children
+    // Blocks until collective is completed
+    void ExecuteCollectives();
+
+    // Perform results validation - compare output to expected
+    void ValidateResults(bool& isCorrect, int collId = -1, int const rank = -1);
+
+    // Release allocated memory
+    void DeallocateMem(int collId = -1, int const rank = -1);
```
Comment (akolliasAMD) [inline test/common/TestBed.hpp:81]: might make sense to be grouped with AllocateMem Function

Comment (gilbertlee-amd) [reply test/common/TestBed.hpp:81]: Currently this is ordered in the general flow through a collective call.  Deallocation of memory generally occurs after results have been validated.

File: test/common/PtrUnion.cpp (hunk 5 above)
Comment (gilbertlee-amd) [reply test/common/PtrUnion.cpp:51]: Unfortunately, this wouldn't provide information about the number of byes

**Review by wenkaidu — APPROVED** [review]

**Review by edgargabriel — APPROVED** [review]
Modulo the test results discussed yesterday in the meeting, I am overall fine with the it.

**Pull request merged**